src-electron/generator/matter/darwin/Framework/CHIP/templates/helper.js

```diff
@@ -711,6 +734,90 @@ function wasIntroducedBeforeRelease(releaseName, clusterName, options) {
   return data.indexOf(introducedRelease) < data.indexOf(referenceRelease);
 }
 
+function wasRemoved(cluster, options) {
+  const data = fetchAvailabilityData(this.global);
+  const path = makeAvailabilityPath(cluster, options);
+
+  let removedRelease = findReleaseForPath(data, ['removed', ...path], options);
+  return removedRelease !== undefined;
+}
+
+function and() {
+  let args = [...arguments];
+  // Strip off the options arg.
+  let options = args.pop();
+
+  return args.reduce((running, current) => {
+    if (current instanceof Promise) {
+      throw new Error(
+        "Promise passed as argument to 'and'.  You probably want to use async_if/async_and"
+      );
+    }
+    return running && current;
+  }, true);
+}
+
+function or() {
+  let args = [...arguments];
+  // Strip off the options arg.
+  let options = args.pop();
+
+  return args.reduce((running, current) => {
+    if (current instanceof Promise) {
+      throw new Error(
+        "Promise passed as argument to 'or'.  You probably want to use async_if/async_or"
+      );
+    }
+    return running || current;
+  }, false);
+}
+
+function not(value) {
+  if (value instanceof Promise) {
+    throw new Error(
+      "Promise passed as argument to 'not'.  You probably want to use async_if/async_not"
```

Comment:
Could you not simply have one `not` and then it detects if it's a promise and does what `async_not` does?

Is there a value of explicitely stating whether you need a promise or not, compared to dynamically reacting on either a promise or not a promise being passed in?

Reply:
I could have done that, but:

1. Then I could not use `async function`, because that will _always_ return a promise.  I would have had to do manual promise chaining and either return a promise or a boolean.
2. Use of `not` with a promise in an `if` would silently do the wrong thing.

So I went with something being a little more explicit: if you see `if` being mixed with `async_not` you know you are doing it wrong, and if you see `if` with `not` then any failures due to promises sneaking in will be not be silent.  And the code is cleaner due to being able to use `async function` for the async version.